Add the MergeSort

Sunbelife opened the pull request:
I have added the mergesort and its benchmark to our project, they also passed google test.

Signed-off-by: Sunbelife Sunbelife@foxmail.com

Nniy committed:
Add Single Random Data Generator

Still working on new features.

Signed-off-by: Zongyu <zy728@york.ac.uk>

Nniy committed:
Added Data Comparing Feature

Added data comparing feature, but still need to be optimized.

Signed-off-by: Zongyu <zy728@york.ac.uk>

0xsuu committed:
Merge branch 'master' into feature/SingleData

And fix conflict.

Signed-off-by: al1ensuu <462993145@qq.com>

Nniy committed:
Merge From master

Merge from master

Signed-off-by: Zongyu <zy728@york.ac.uk>

Nniy committed:
Fix Conflicts

Merge remote-tracking branch 'origin/feature/SingleData' into feature/SingleData

Signed-off-by: Zongyu <zy728@york.ac.uk>

0xsuu committed:
Merge remote-tracking branch 'origin/master' into feature/SingleData

0xsuu committed:
Update comparison with single datum generator

Signed-off-by: al1ensuu <462993145@qq.com>

BlueCocoa committed:
A few improvements

- Remove redundant include in DataGenerator.cpp since we include
headers in DataGenerator.h
- Using generateRandomDataFromSet as a wrapper for
generateRandomData<std::string>
- Using macro DATA_GENERATOR_DEFAULT_COUNT
- Using macro DATA_GENERATOR_DEFAULT_STRING_MAX_LENGTH
- Make all methods class method (corresponding changes already applied
to example file)

0xsuu committed:
Merge remote-tracking branch 'origin/feature/SingleData' into feature/SingleData

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Merge branch 'feature/SingleData' of https://github.com/Llama-Group/Project-Llama into feature/SingleData

BlueCocoa committed:
Align code

review-ninja commented:
<a href="https://app.review.ninja/Llama-Group/Project-Llama/pull/127" target="_blank"><img src="https://app.review.ninja/63898156/pull/127/badge" alt="ReviewNinja"/>

0xsuu commented on llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp:
```diff
@@ -15,17 +15,18 @@
 //  limitations under the License.
```
Dont change the format before, we use 4 spaces! not 2

0xsuu commented:
ＤＯ ＮＯＴ ＣＨＡＮＧＥ ＵＮＲＥＬＡＴＥＤ ＦＩＬＥＳ！

Nniy committed:
Debug SingleData

Debug singleData.

Signed-off-by: Zongyu <zy728@york.ac.uk>

Nniy committed:
Add Exceptions

Added exceptions to warning if the the given data is already maximum or minimum.

Signed-off-by: Zongyu <zy728@york.ac.uk>

Nniy committed:
Fix Bugs

Fixed bugs.

Signed-off-by: Zongyu <zy728@york.ac.uk>

0xsuu committed:
Fix Alex's super stupid bug

Don't assign int with double, hahahhahahahahahahah

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Fix some nits

Causing lint error and travis fail to build.

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu commented on llama_cpp/src/Sort/MergeSort.h:
```diff
@@ -0,0 +1,79 @@
+        std::swap(tmp.at(k++), arr->at(j++));
+    }
+
+    while (i <= mid)
```
Must use {

0xsuu commented on llama_cpp/src/Sort/MergeSort.h:
```diff
@@ -0,0 +1,79 @@
+
+  void doSort(std::vector<SortObject<T>> *reformedArray) override {
+    mergesort(reformedArray, 0, reformedArray->size() - 1);
+  }
```
Four spaces

0xsuu commented on llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp:
```diff
@@ -40,289 +41,317 @@ using llama::Sort;
 struct SortObjectTest : public ::testing::Test {
-    SortObject<int> objInt1;
-    SortObject<int> objInt2;
+  SortObject<int> objInt1;
```
Four spaces

0xsuu commented:
We fucking use 4 spaces not 2! Still, don't change unrelated files.

0xsuu commented on llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp:
```diff
@@ -208,7 +229,7 @@ TEST_F(SortCorrectnessTest, BruteForceRandomStringsCorrectness) {
 }
 
-// Testing for Insertion Sort.
+/* Testing for Insertion Sort. */
```
Don't change this.

0xsuu replied:
Use //

0xsuu commented on llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp:
```diff
@@ -167,8 +183,10 @@ struct SortCorrectnessTest : public ::testing::Test {
+        mContinuousData.generateContinuousData(&continuousDoublesReversed, 10,
+                                                                                     true);
+        mContinuousData.generateContinuousData(&continuousStringsReversed, 10,
+                                                                                     true);
```
WTF are these.

0xsuu committed:
Add NE for Sort

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Merge remote-tracking branch 'origin/master' into feature/SingleData

0xsuu commented on llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp:
```diff
@@ -159,6 +160,10 @@ struct SortCorrectnessTest : public ::testing::Test {
     llama::SelectionSort<double> doubleSe;
     llama::SelectionSort<string> stringSe;
 
+    llama::MergeSort<int> intme;
```
intMe

0xsuu committed:
Solving "Invalid CMake command" issue

0xsuu committed:
Refactoring utilut add DG UT

We are doing too much fucking things on this branch! Should ignore coverall coverage!

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Add NE test and use static methods of DG

Signed-off-by: al1ensuu <462993145@qq.com>

Sunbelife committed:
change "//" to "/* */"

in SortUnitTest.cpp.

Signed-off-by: Sunbelife <Sunbelife@foxmail.com>

0xsuu committed:
Add exception test

Whooray! go to sleep now! leave shitty thing to Alex!

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Add minor massive to speed up Sort test

Cuz we have so many dumbass sorts!

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Style fixes

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Remove coverage for UT

Signed-off-by: al1ensuu <462993145@qq.com>

BlueCocoa committed:
Merge pull request #128 from Llama-Group/fix/solve-invalid-cmake-command

Solving "Invalid CMake command" issue

0xsuu committed:
Merge pull request #108 from Llama-Group/feature/SingleData

Single Random Data Generator

Nniy committed:
Add Random Bool Generator

Added random bool generator and optimise some functions.

Signed-off-by: Zongyu <zy728@york.ac.uk>

BlueCocoa committed:
The same as #77

The same as #77
Headers at /llama_cpp/Sources/llama/Headers in Project Navigator

Signed-off-by: 0xBBC <0xbbc@0xbbc.com>

0xsuu committed:
Merge pull request #132 from Llama-Group/fix/xcode-not-showing-headers-oh-my-god-its-already-used

The same as #77

0xsuu committed:
Add test to cover bool array generation

Signed-off-by: al1ensuu <462993145@qq.com>

BlueCocoa committed:
Merge pull request #131 from Llama-Group/feature/RandomBool

Add Random Bool Generator

BlueCocoa committed:
Lots of changes made, please see the list below

#### Support for general integral number and float point number
- generateRandomData
- generateContinuousData
- generateSingleDatum
- generateRandomDataFromRange

#### No more template specialization for std::string
No template if anything related to std::string. Since we always write
template specialization for std::string, why not make they independent
from the template?

#### Others
Corresponding changes already applied to examples and unit tests.

Signed-off-by: 0xBBC <0xbbc@0xbbc.com>

BlueCocoa committed:
Add stdexcept in header file

BlueCocoa committed:
Generate random bool

0xsuu committed:
Fix arithmetic rd error

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
A final fix

Hooray! Preparing to merge.

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Fixing nits

Signed-off-by: al1ensuu <462993145@qq.com>

BlueCocoa committed:
Merge pull request #134 from Llama-Group/feature/general-data-generator

DataGenerator Refactoring and Recontructing

BlueCocoa committed:
Type alias & __unused

- Using type alias instead of macro for ```TYPE_INTEGRAL``` and
```TYPE_FLOATING_POINT```.
- And ```__unused``` qualified to some arguments in case our user get
confused.

BlueCocoa committed:
Oops, __unused is not defined in gcc

0xsuu committed:
Merge pull request #137 from Llama-Group/improve/data_generator

Type alias & __unused

BlueCocoa committed:
Add Logger and corresponding examples

- Disable [build/c++11] of lintcpp. Because we’re using some C++14
features now.
- Add module_check

Signed-off-by: 0xBBC <0xbbc@0xbbc.com>

BlueCocoa committed:
Add mutex header

BlueCocoa committed:
Add algorithm header

BlueCocoa committed:
Find and link installed thread model

BlueCocoa committed:
Merge pull request #140 from Llama-Group/feature/Logger

Add Logger and corresponding examples

BlueCocoa committed:
Add unit test for Logger

BlueCocoa committed:
Change the name of the unit test of Logger, and adjust the timing of calling flush in error_x.

BlueCocoa committed:
Remove operator << overloading in Logger

Given current API design, there is no need to overload operator <<.

Signed-off-by: 0xBBC <0xbbc@0xbbc.com>

0xsuu committed:
Merge pull request #141 from Llama-Group/feature/Logger

Add unit test for Logger

BlueCocoa committed:
Add LLAMA_DEPRECATED macro

Provide a transitional period for users if there is or going to be any
API change.

0xsuu committed:
Merge pull request #142 from Llama-Group/feature/deprecated

Add LLAMA_DEPRECATED macro

BlueCocoa committed:
Fix some typo and attribute((noreturn))

BlueCocoa committed:
Merge pull request #143 from Llama-Group/fix/attribute-and-typo-for-logger

Fix some typo and attribute((noreturn))

Sunbelife committed:
Merge branch 'feature/add-merge-sort' of https://github.com/Llama-Group/Project-Llama into origin/feature/add-merge-sort

0xsuu commented on llama_cpp/benchmark/sort_benchmark/sort_benchmark.cpp:
```diff
@@ -198,6 +224,15 @@ static void SeSortClass(benchmark::State& state) {
 }
 
+static void MeSortClass(benchmark::State& state) {
+    llama::MergeSort<CustomisedClass> MeSort = llama::MergeSort<CustomisedClass>();
```
meSort not MeSort

0xsuu commented on llama_cpp/benchmark/sort_benchmark/sort_benchmark.cpp:
```diff
@@ -198,6 +224,15 @@ static void SeSortClass(benchmark::State& state) {
+    std::vector<CustomisedClass> a = {CustomisedClass(5), CustomisedClass(4),
+                          CustomisedClass(3), CustomisedClass(2), CustomisedClass(1)};
+    while (state.KeepRunning()) {
+        MeSort.performSort(&a);
```
meSort not MeSort

0xsuu commented on llama_cpp/src/Sort/MergeSort.h:
```diff
@@ -0,0 +1,80 @@
+    int i, j, k;
+    i = low;
+    j = mid + 1;
+    k = 0;
```
int i = low; not int i; i = low;
Same for j, k

0xsuu commented on llama_cpp/src/Sort/MergeSort.h:
```diff
@@ -0,0 +1,80 @@
+namespace llama {
+template <class T> class MergeSort : public Sort<T> {
+public:
+  MergeSort() {}
```
***\* 4 ***\* spaces not 2

0xsuu replied:
Change the whole file back to 4 spaces. DO NOT USE FIND & REPLACE.

0xsuu commented on llama_cpp/src/Sort/MergeSort.h:
```diff
@@ -0,0 +1,80 @@
+    int mid;
+    if (low < high) {
+      mid = (low + high) / 2;
+      // left hand
```
Left hand side.

0xsuu commented on llama_cpp/src/Sort/MergeSort.h:
```diff
@@ -0,0 +1,80 @@
+      mid = (low + high) / 2;
+      // left hand
+      mergesort(arr, low, mid);
+      // right hand
```
Right hand side.

0xsuu commented on llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp:
```diff
@@ -192,7 +197,7 @@ struct SortCorrectnessTest : public ::testing::Test {
 };
 
-// Testing for Brute Force Sort.
+/* Testing for Brute Force Sort. */
```
Use //

0xsuu commented on llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp:
```diff
@@ -224,7 +229,7 @@ TEST_F(SortCorrectnessTest, InsertionRandomStringsCorrectness) {
 }
 
-// Testing for Bubble Sort.
+/* Testing for Bubble Sort. */
```
Use //

0xsuu commented on llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp:
```diff
@@ -240,7 +245,7 @@ TEST_F(SortCorrectnessTest, BubbleRandomStringsCorrectness) {
 }
 
-// Testing for Selection Sort.
+/* Testing for Selection Sort. */
```
Use //

0xsuu commented on llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp:
```diff
@@ -256,6 +261,21 @@ TEST_F(SortCorrectnessTest, SelectionRandomStringsCorrectness) {
     EXPECT_EQ(randomStringsCorrect, randomStrings);
 }
 
+/* Testing for Merge Sort. */
```
Use //

Sunbelife committed:
Fixed indentation problem

I fixed the indentation problem in MergeSort.cpp and some comment problem.

Signed-off-by: Sunbelife <Sunbelife@foxmail.com>
